### PR TITLE
add mechanics for retrieving filter properties.

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will add mechanics for getting filter properties

### CODE REVIEW

**Comment by phlptp** [comment]
will add the C version of this soon to this PR

**Comment by phlptp** [comment]
Fixes #2669

**Comment by trevorhardy** [comment]
@phlptp are you wanting a review on this?

**Comment by phlptp** [comment]
> @phlptp are you wanting a review on this?

Yes, Mainly to make sure it meets your needs regarding #2669

File: src/helics/application_api/FilterOperations.cpp
Comment (josephmckinsey) [inline src/helics/application_api/FilterOperations.cpp:1]: I'm not a big fan of reusing two doubles named param1 and param2, and this definitely goes against the spirit of the linter here. Any way we can use a tagged union or std::variant instead? Might be helpful in the if statements too.

File: src/helics/application_api/FilterOperations.cpp [hunk 1]
```diff
@@ -273,13 +350,45 @@ void RerouteFilterOperation::setString(std::string_view property, std::string_vi
         }
         catch (const std::regex_error& re) {
             std::cerr << "filter expression is not a valid Regular expression " << re.what()
-                      << std::endl;
+                      << '\n';
```
Comment (josephmckinsey) [inline src/helics/application_api/FilterOperations.cpp:353]: Is this to avoid flushing `cerr`?

**Review by josephmckinsey — COMMENTED** [review]
It would be nice to return more meaningful errors.

File: src/helics/application_api/FilterOperations.cpp (hunk 1 above)
Comment (phlptp) [reply src/helics/application_api/FilterOperations.cpp:353]: I think cerr is flushed anyway, so I don't think the flushing matters here.

File: src/helics/application_api/FilterOperations.cpp [hunk 2]
```diff
@@ -23,9 +23,19 @@ SPDX-License-Identifier: BSD-3-Clause
 #include <vector>
 
 namespace helics {
+
 void FilterOperations::set(std::string_view /*property*/, double /*val*/) {}
 void FilterOperations::setString(std::string_view /*property*/, std::string_view /*val*/) {}
 
+double FilterOperations::getProperty(std::string_view /*property*/)
+{
+    return invalidDouble;
```
[the remark this answers is not in the document]
Comment (phlptp) [reply src/helics/application_api/FilterOperations.cpp:32]: Then we have to do a bunch of error checking through a lot of calls.

File: src/helics/application_api/FilterOperations.cpp
Comment (phlptp) [reply src/helics/application_api/FilterOperations.cpp:1]: the parameters have different meanings depending on which random distribution is used, so there is no standard naming convention there.

**Comment by trevorhardy** [comment]
@josephmckinsey, can you get the C-based additions @phlptp just added into PyHELICS for testing? I don't know if we want to use the libclang script or do it manually. Maybe @nightlark has an idea on the best way to do this?

**Comment by nightlark** [comment]
For pyhelics changes: use https://github.com/GMLC-TDC/cheader2json#usage to dump ast.json files for the previous helics release (using the helics.h backup header file) and for the new development version of helics (using the new helics.h backup header file), then use it to do a diff of the ast.json files to get the changes that need to be made to pyhelics.

File: src/helics/application_api/FilterOperations.hpp [hunk 3]
```diff
@@ -51,6 +51,19 @@ class FilterOperations {
     @param val the numerical value of the property
     */
     virtual void setString(std::string_view property, std::string_view val);
+
+    /** get a property from a filter
+    @param property the name of the property of the filter to change
+    @param val the numerical value of the property
```
Comment (nightlark) [inline src/helics/application_api/FilterOperations.hpp:57]:
```suggestion
```

File: src/helics/application_api/FilterOperations.hpp [hunk 4]
```diff
@@ -51,6 +51,19 @@ class FilterOperations {
     @param val the numerical value of the property
     */
     virtual void setString(std::string_view property, std::string_view val);
+
+    /** get a property from a filter
+    @param property the name of the property of the filter to change
```
Comment (nightlark) [inline src/helics/application_api/FilterOperations.hpp:56]:
```suggestion
    @param property the name of the property of the filter to get
```

File: src/helics/application_api/FilterOperations.hpp [hunk 5]
```diff
@@ -51,6 +51,19 @@ class FilterOperations {
...
+    @param val the numerical value of the property
+    */
+    virtual double getProperty(std::string_view property);
+    /** get a string property on a filter
+    @param property the name of the property of the filter to change
```
Comment (nightlark) [inline src/helics/application_api/FilterOperations.hpp:61]:
```suggestion
    @param property the name of the property of the filter to get
```

File: src/helics/application_api/FilterOperations.hpp [hunk 6]
```diff
@@ -51,6 +51,19 @@ class FilterOperations {
...
+    @param val the numerical value of the property
```
Comment (nightlark) [inline src/helics/application_api/FilterOperations.hpp:62]:
```suggestion
```

File: src/helics/application_api/Filters.hpp [hunk 7]
```diff
@@ -84,6 +84,17 @@ class HELICS_CXX_EXPORT Filter: public Interface {
     */
     virtual void setString(std::string_view property, std::string_view val);
 
+    /** get a string property on a filter
+    @param property the name of the property of the filter to change
```
Comment (nightlark) [inline src/helics/application_api/Filters.hpp:88]:
```suggestion
    @param property the name of the property of the filter to get
```

File: src/helics/application_api/Filters.hpp [hunk 8]
```diff
@@ -84,6 +84,17 @@ class HELICS_CXX_EXPORT Filter: public Interface {
...
+    */
+    virtual std::string getString(std::string_view property);
+
+    /** set a double valued property on a filter
```
Comment (nightlark) [inline src/helics/application_api/Filters.hpp:92]:
```suggestion
    /** get a double valued property of a filter
```

File: src/helics/cpp98/Filter.hpp [hunk 9]
```diff
@@ -54,6 +54,20 @@ class Filter {
     {
         helicsFilterSetString(filt, property.c_str(), val.c_str(), hThrowOnError());
     }
+    /** get the value of a filter property
+    @param property the name of the property of the filter to change
```
Comment (nightlark) [inline src/helics/cpp98/Filter.hpp:58]:
```suggestion
    @param property the name of the property of the filter to get
```

File: src/helics/cpp98/Filter.hpp [hunk 10]
```diff
@@ -54,6 +54,20 @@ class Filter {
...
+    */
+    double getProperty(const std::string& property)
+    {
+        return helicsFilterGetPropertyDouble(filt, property.c_str());
+    }
+    /** get a string property on a filter
+    @param property the name of the property of the filter to change
```
Comment (nightlark) [inline src/helics/cpp98/Filter.hpp:65]:
```suggestion
    @param property the name of the property of the filter to get
```

File: src/helics/shared_api_library/MessageFilters.h [hunk 11]
```diff
@@ -198,6 +198,28 @@ HELICS_EXPORT void helicsFilterSet(HelicsFilter filt, const char* prop, double v
  */
 HELICS_EXPORT void helicsFilterSetString(HelicsFilter filt, const char* prop, const char* val, HelicsError* err);
 
+/**
+* Get a double property from a filter
```
Comment (nightlark) [inline src/helics/shared_api_library/MessageFilters.h:202]:
```suggestion
* Get a double property from a filter.
```

File: src/helics/shared_api_library/MessageFilters.h [hunk 12]
```diff
@@ -198,6 +198,28 @@ HELICS_EXPORT void helicsFilterSet(HelicsFilter filt, const char* prop, double v
...
+*
+* @param filt The filter to modify.
```
Comment (nightlark) [inline src/helics/shared_api_library/MessageFilters.h:204]:
```suggestion
* @param filt The filter to retrieve a value from.
```

File: src/helics/shared_api_library/MessageFilters.h [hunk 13]
```diff
@@ -198,6 +198,28 @@ HELICS_EXPORT void helicsFilterSet(HelicsFilter filt, const char* prop, double v
...
+* @param prop A string containing the property to set.
```
Comment (nightlark) [inline src/helics/shared_api_library/MessageFilters.h:205]:
```suggestion
* @param prop A string containing the property to get.
```

File: src/helics/shared_api_library/MessageFilters.h [hunk 14]
```diff
@@ -198,6 +198,28 @@ HELICS_EXPORT void helicsFilterSet(HelicsFilter filt, const char* prop, double v
...
+* @param[in,out] err A pointer to an error object for catching errors.
+
+*/
+HELICS_EXPORT double helicsFilterGetPropertyDouble(HelicsFilter filt, const char* prop);
+
+/**
+* Set a string property on a filter.  The string output memory is valid until a subsequent call to to getPropertyString on the particular
+filter
+*
+* @param filt The filter to modify.
```
Comment (nightlark) [inline src/helics/shared_api_library/MessageFilters.h:216]:
```suggestion
* @param filt The filter to retrieve a value from.
```

File: src/helics/shared_api_library/MessageFilters.h [hunk 15]
```diff
@@ -198,6 +198,28 @@ HELICS_EXPORT void helicsFilterSet(HelicsFilter filt, const char* prop, double v
...
+* @param prop A string containing the property to set.
```
Comment (nightlark) [inline src/helics/shared_api_library/MessageFilters.h:217]:
```suggestion
* @param prop A string containing the property to get.
```

File: src/helics/shared_api_library/MessageFilters.h [hunk 16]
```diff
@@ -198,6 +198,28 @@ HELICS_EXPORT void helicsFilterSet(HelicsFilter filt, const char* prop, double v
...
+*
+* @param[in,out] err A pointer to an error object for catching errors.
+
+*/
+HELICS_EXPORT double helicsFilterGetPropertyDouble(HelicsFilter filt, const char* prop);
+
+/**
+* Set a string property on a filter.  The string output memory is valid until a subsequent call to to getPropertyString on the particular
```
Comment (nightlark) [inline src/helics/shared_api_library/MessageFilters.h:213]:
```suggestion
* Get a string property on a filter.  The string output memory is valid until a subsequent call to getPropertyString on the particular
```

File: src/helics/shared_api_library/MessageFilters.h [hunk 17]
```diff
@@ -198,6 +198,28 @@ HELICS_EXPORT void helicsFilterSet(HelicsFilter filt, const char* prop, double v
...
+filter
```
Comment (nightlark) [inline src/helics/shared_api_library/MessageFilters.h:214]:
```suggestion
filter.
```

**Review by nightlark — APPROVED** [review]

**Comment by nightlark** [comment]
> > For pyhelics changes: use https://github.com/GMLC-TDC/cheader2json#usage to dump ast.json files for the previous helics release (using the helics.h backup header file)...
> 
> Any update on getting these changes into PyHELICS for testing?

I don't think anyone has been working on this.

**Pull request merged**